### PR TITLE
Add initial files

### PR DESCRIPTION
**Pull request by abkieling**
Initial implementation of the XML - JSON transformation plugin. I tried to write unit tests for the transformers but had issues asserting the results due to the fact that the order of JSON object fields isn't deterministic. Please review the code and the configuration page. I hope it's OK.

### CODE REVIEW

**Comment by msavy** [comment]
Wonder if we can use something like this as a test-only dep
http://jsonassert.skyscreamer.org/ - it solves the issue you describe

**Comment by abkieling** [comment]
Thanks for the tip! I'll give it a try tonight.

**Comment by EricWittmann** [comment]
I'd actually prefer using jackson to solve this.  Brandon did this very thing to solve the non-deterministic json property issue in the services test:

https://github.com/apiman/apiman/blob/master/manager/test/api/src/test/java/io/apiman/manager/test/ServicesTest.java#L53-L56

I'd prefer using jackson because I believe it's already a dependency being pulled in.

File: transformation-policy/pom.xml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    <relativePath>../pom.xml</relativePath>
+  </parent>
+  <artifactId>apiman-plugins-transformation-policy</artifactId>
+  <packaging>war</packaging>
+  <name>apiman-plugins-transformation-policy</name>
+  <dependencies>
+    <dependency>
+      <groupId>org.json</groupId>
+      <artifactId>json</artifactId>
+      <version>20140107</version>
```
Comment (EricWittmann) [inline transformation-policy/pom.xml]: Can you make sure all dependency versions are managed via the parent pom's <dependencyManagement> section?  We should not have any dependency versions in the individual plugin poms.

File: transformation-policy/pom.xml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    </dependency>
+    <!-- apiman dependencies (must be excluded from the WAR) -->
+    <dependency>
+      <groupId>io.apiman</groupId>
+      <artifactId>apiman-gateway-engine-core</artifactId>
+      <version>1.1.4-SNAPSHOT</version>
```
Comment (EricWittmann) [inline transformation-policy/pom.xml]: I believe the apiman project versions are already managed in the parent pom - so you should be able to simply remove these...

File: transformation-policy/src/test/java/io/apiman/plugins/transformation_policy/TransformationPolicyTest.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        
+        PolicyTestResponse response = send(request);
+        
+        String expectedResponse = "<root><name>apiman</name></root>";
+        assertEquals("application/xml", response.header("Content-Type"));
+        assertNull(response.header("Content-Length"));
+        assertEquals(expectedResponse, response.body());
+    }
+
+    public static class ProduceJsonBackEndService implements IPolicyTestBackEndService {
```
Comment (EricWittmann) [inline transformation-policy/src/test/java/io/apiman/plugins/transformation_policy/TransformationPolicyTest.java]: Can you break out the back-end service mock classes out into separate java files?  No big deal, just a personal preference. :)

**Comment by EricWittmann** [comment]
This looks really nice, thanks Alexandre. :)

Could you also add at least one test that uses more complex data?  Something with a more substantial payload with nested/child values and whatnot.  Examples:

http://www.xmlfiles.com/examples/plant_catalog.xml
http://www.xmlfiles.com/examples/simple.xml

**Comment by msavy** [comment]
@EricWittmann my argument for jsonassert is that, as I understand it, the jackson comparison method used by Brandon is still prone to certain non-determinism failures you can't easily legislate for. For instance, if you want loose array ordering ['a', 'b'] == ['b', 'a'], you'll end up with problems.
(See also: http://stackoverflow.com/a/10047161/2766538)

JSONAssert itself could presumably be used as a test-only dep, and itself only has 2 dependencies.

**Comment by EricWittmann** [comment]
I'm not against using jsonassert if it's needed.  But if jackson's compare is sufficient, and it's already available..... :)

**Comment by abkieling** [comment]
Eric, remember you made some changes to the plugin test classes a couple of weeks ago? I need them for my tests to run, but they are in apiman 1.1.4-SNAPSHOT. The parent pom references 1.1.3.Final. Can I change the parent pom?

**Comment by EricWittmann** [comment]
Yes go ahead and change the parent pom to 1.1.4-SNAPSHOT.  I will deploy 1.1.4-SNAPSHOT into jboss nexus right now.

**Comment by abkieling** [comment]
Jackson's ObjectMapper is working well to compare JSON data. I have tried to use the xmlunit library to compare XML data because it's already present as a dependency. It works, but it requires the XMLs to be formatted exactly the same way. I did some testing with Jackson's XmlMapper and it worked better. The only problem is that it's a new dependency (jackson-dataformat-xml). Any suggestions?

**Comment by EricWittmann** [comment]
Maybe take one more swing at using xmlunit.  I am pretty sure you can tell xmlunit to ignore certain classes of difference, allowing you to assert that the xml is "similar".

http://xmlunit.sourceforge.net/userguide/html/index.html#Configuring%20XMLUnit

For example you could use a custom DifferenceListener.  

That said - if it's a pain, I'm not against bringing in jackson-dataformat-xml as a dependency.

**Comment by abkieling** [comment]
Thanks for asking me to persevere. I've found a way to configure xmlunit to ignore white spaces and line breaks. So there are no new dependencies to test the plugin.

**Comment by EricWittmann** [comment]
Very nice.  Is this PR ready to be merged now?  It looks pretty good to me!

**Comment by abkieling** [comment]
While doing more testing, I have found that something goes wrong when processing the content of responses with Content-Encoding: gzip. It doesn't seem to be specific to this plugin because the jsonp plugin also fails in that situation. Here is a URL that can be tested http://api.brewerydb.com/v2/

**Comment by EricWittmann** [comment]
If the content is encoded, then we would need a data policy that unpacked the content on the fly.  That is out of scope for this PR, of course.  But it's a good thought and we should create a JIRA issue for it.  Would you like to do that or shall I?

**Comment by msavy** [comment]
@EricWittmann this is a really interesting issue. Generally I can't imagine any policy wanting to inspect the body whilst still compressed, but we really don't want to bother doing the decompression unless we know something is going to touch the body.

I should look and see what Vert.x does, as I'm not sure whether it automatically decompressed.

Perhaps a nice solution, rather than requiring a decompression policy, could be to check whether we have any policies lined up which are `instanceof IDataPolicy`, then (if not already decompressed by the underlying layer) perform the decompression.

Then it asks the interesting question as to whether we should recompress to the back-end (endpoint setting?).

**Comment by EricWittmann** [comment]
@msavy - you make a good point.  It would be nicer for users if the platform simply handled it for you.  Can you think of any use-cases where a policy would want to have access to the compressed data stream?  I can't.

So yeah I think you're right - we can make it a core feature of the system.  Decompress the content only if there is a data policy in the chain.  Not sure about whether to compress it again on the way out.  I suppose we should do that automatically as well, but only if we auto-decompressed it. :)

**Comment by abkieling** [comment]
I'd be happy to fix this issue. Just need some directions after @msavy looks into the Vert.x code.

**Comment by EricWittmann** [comment]
@alexkieling - this PR looks pretty good, nice testing.  Is it OK to merge it?

**Comment by EricWittmann** [comment]
We'll handle the gzip issue in a separate JIRA and separate PR.

**Comment by abkieling** [comment]
It's OK to merge. I think it's good as a first version. We can improve it based on user feedback.

**Pull request merged**